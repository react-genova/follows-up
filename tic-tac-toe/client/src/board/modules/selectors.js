import { createSelector } from 'reselect';

let getBoardRoot = state => state;

export const setBoardRoot = (getter) => {
    getBoardRoot = getter;
};

export const getBoard = state => getBoardRoot(state);

export const getBoardValues = createSelector([getBoard], board => board.get('values').toJS());
