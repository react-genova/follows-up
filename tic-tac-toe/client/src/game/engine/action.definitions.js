export const BEGIN_GAME = 'BEGIN_GAME';
export const END_GAME = 'END_GAME';
export const EXIT_GAME = 'EXIT_GAME';
