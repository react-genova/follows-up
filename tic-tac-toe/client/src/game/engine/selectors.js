import { createSelector } from 'reselect';
import { IDLE } from './types/game.status.constants';

let getEngineRoot = state => state;

export const setEngineRoot = (getter) => {
    getEngineRoot = getter;
};

const getEngine = state => getEngineRoot(state);

const getEngineHistory = createSelector([getEngine], engine => engine.get('history'));

export const getGameStatus = createSelector([getEngine], engine => engine.get('status'));

export const getLastResult = createSelector([getEngine], engine => engine.get('result'));

export const getTotalDraws = createSelector([getEngineHistory], engine => engine.get('draws'));

export const getTotalPlayer1Victories = createSelector([getEngineHistory], engine => engine.get('player1victories'));

export const getTotalPlayer2Victories = createSelector([getEngineHistory], engine => engine.get('player2victories'));

export const isGameStarted = createSelector([getGameStatus], status => IDLE !== status);
