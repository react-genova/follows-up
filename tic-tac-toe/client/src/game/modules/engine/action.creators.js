import { BEGIN_GAME, END_GAME, EXIT_GAME } from './action.definitions';

export const beginGame = () => ({
    type: BEGIN_GAME,
    payload: { },
});

export const endGame = gameresult => ({
    type: END_GAME,
    payload: {
        gameresult,
    },
});

export const exitGame = () => ({
    type: EXIT_GAME,
    payload: { },
});
