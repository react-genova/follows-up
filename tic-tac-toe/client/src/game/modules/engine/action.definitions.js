export const BEGIN_GAME = 'BEGIN_GAME@game';
export const END_GAME = 'END_GAME@game';
export const EXIT_GAME = 'EXIT_GAME@game';
