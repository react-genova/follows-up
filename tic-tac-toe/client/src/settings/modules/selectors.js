import { createSelector } from 'reselect';

let getSettingsRoot = state => state;

export const setSettingsRoot = (getter) => {
    getSettingsRoot = getter;
};

const getSettings = state => getSettingsRoot(state);

const getPlayer1 = createSelector([getSettings], settings => settings.get('player1'));

const getPlayer2 = createSelector([getSettings], settings => settings.get('player2'));

const getOptions = createSelector([getSettings], settings => settings.get('options'));

export const getPlayer1Name = createSelector([getPlayer1], player1 => player1.get('name'));

export const getPlayer1Symbol = createSelector([getPlayer1], player1 => player1.get('symbol'));

export const getPlayer1Ready = createSelector([getPlayer1], player1 => player1.get('ready'));

export const getPlayer1Type = createSelector([getPlayer1], player1 => player1.get('type'));

export const getPlayer2Name = createSelector([getPlayer2], player2 => player2.get('name'));

export const getPlayer2Symbol = createSelector([getPlayer2], player2 => player2.get('symbol'));

export const getPlayer2Ready = createSelector([getPlayer2], player2 => player2.get('ready'));

export const getPlayer2Type = createSelector([getPlayer2], player2 => player2.get('type'));

export const getAutoplay = createSelector([getOptions], options => options.get('autoplay'));

export const getBotSpeed = createSelector([getOptions], options => options.get('botSpeed'));
