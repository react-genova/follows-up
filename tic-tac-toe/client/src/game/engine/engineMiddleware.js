import {
    getPlayer1Ready, getPlayer2Ready, getPlayer1Symbol,
} from '../../settings/modules/selectors';
import { isGameStarted, isGameIdle } from './selectors';
import { beginGame, endGame } from './action.creators';
import { getBoardValues } from '../../board/modules/selectors';
import getGameResults from './engine.utils';
import { DRAW, PLAYER_1_WON, PLAYER_2_WON } from './types/game.results.constants';

const engineMiddleware = ({ getState, dispatch }) => next => (action) => {
    const resultAction = next(action);
    const state = getState();
    if (isGameStarted(state)) {
        const result = getGameResults(getBoardValues(state));
        if (result.ended) {
            if (result.draw) {
                dispatch(endGame(DRAW));
            } else if (result.winner === getPlayer1Symbol(state)) {
                dispatch(endGame(PLAYER_1_WON));
            } else {
                dispatch(endGame(PLAYER_2_WON));
            }
        }
    } else if (isGameIdle(state) && getPlayer1Ready(state) && getPlayer2Ready(state)) {
        dispatch(beginGame());
    }
    return resultAction;
};

export default engineMiddleware;
