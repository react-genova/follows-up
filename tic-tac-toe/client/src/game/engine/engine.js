import { initialEngineState } from './types/engine.types';
import { BEGIN_GAME, END_GAME, EXIT_GAME } from './action.definitions';
import {
    DRAW, PLAYER_1_WON, PLAYER_2_WON, NONE,
} from './types/game.results.constants';
import { STARTED, WAITING_FOR_START } from './types/game.status.constants';

const getFieldFromResult = gameresult => ({
    [DRAW]: 'draws',
    [PLAYER_1_WON]: 'player1victories',
    [PLAYER_2_WON]: 'player2victories',
}[gameresult]);

const engine = (state = initialEngineState, action = {}) => {
    const { type, payload } = action;
    switch (type) {
    case BEGIN_GAME: return state.withMutations((mutable) => {
        mutable.set('status', STARTED);
        mutable.set('result', NONE);
    });
    case END_GAME: return state.withMutations((mutable) => {
        const { gameresult } = payload;
        mutable.updateIn(['history', getFieldFromResult(gameresult)], value => value + 1);
        mutable.set('status', WAITING_FOR_START);
        mutable.set('result', gameresult);
    });
    case EXIT_GAME: return initialEngineState;
    default: return state;
    }
};

export default engine;
