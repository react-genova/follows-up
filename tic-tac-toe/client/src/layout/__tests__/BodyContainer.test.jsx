import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import { isGameIdle } from '../../game/engine/selectors';

jest.mock('../../game/engine/selectors');
jest.mock('../../board/containers/BoardContainer', () => () => 'BOARD');
jest.mock('../../settings/containers/GameMode', () => () => 'GAMEMODE');

const { BodyContainer, mapStatetoProps } = require('../BodyContainer');

describe('BodyContainer component', () => {
    afterEach(() => {
        cleanup();
        jest.clearAllMocks();
    });

    it('renders without crashing', () => {
        const { container } = render(<BodyContainer gameIdle />);
        expect(container.firstChild).toMatchSnapshot();
        expect(container.firstChild.textContent).toBe('GAMEMODE');
    });

    it('renders without crashing (gameStarted)', () => {
        const { container } = render(<BodyContainer />);
        expect(container.firstChild).toMatchSnapshot();
        expect(container.firstChild.firstChild.textContent).toBe('BOARD');
    });

    it('exports valid props throught mapStatetoProps 1', () => {
        const STATE = 'FAKE_STATE';
        isGameIdle.mockImplementation(() => false);
        expect(mapStatetoProps(STATE).gameIdle).toBe(false);
        expect(isGameIdle).toHaveBeenCalledTimes(1);
        expect(isGameIdle).toHaveBeenCalledWith(STATE);
    });

    it('exports valid props throught mapStatetoProps 2', () => {
        const STATE = 'FAKE_STATE';
        isGameIdle.mockImplementation(() => true);
        expect(mapStatetoProps(STATE).gameIdle).toBe(true);
        expect(isGameIdle).toHaveBeenCalledTimes(1);
        expect(isGameIdle).toHaveBeenCalledWith(STATE);
    });
});
