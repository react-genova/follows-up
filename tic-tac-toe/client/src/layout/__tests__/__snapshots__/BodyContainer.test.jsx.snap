// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BodyContainer component renders without crashing (gameStarted) 1`] = `
<div
  class="sc-bdVaJa koRVId"
>
  <div
    style="display: flex; flex-direction: column;"
  >
    HISTORY
    MENU
  </div>
  BOARD
</div>
`;

exports[`BodyContainer component renders without crashing 1`] = `GAMEMODE`;
